server/service/oauth.go

```diff
@@ -88,12 +98,83 @@ func (o *OAuth) GetUser(ctx context.Context, code string) (*model.User, error) {
 	var user githubUser
 	err = json.NewDecoder(resp.Body).Decode(&user)
 	if err != nil {
-		return nil, fmt.Errorf("can't parse github response: %s", err)
+		return nil, fmt.Errorf("can't parse github user response: %s", err)
+	}
+
+	if o.restrictAccess != "" {
+		if err := o.checkAccess(client, o.restrictAccess, user.Login); err != nil {
+			return nil, err
+		}
 	}
+
+	role := model.Requester
+	if o.restrictReviewerAccess != "" {
+		err := o.checkAccess(client, o.restrictReviewerAccess, user.Login)
+		if err != nil && err != NoAccessError {
+			return nil, err
+		}
+		if err == NoAccessError {
+			role = model.Worker
+		}
+	}
+
 	return &model.User{
 		ID:        user.ID,
 		Login:     user.Login,
 		Username:  user.Username,
 		AvatarURL: user.AvatarURL,
+		Role:      role,
 	}, nil
 }
+
+const orgPrefix = "org:"
+const teamPrefix = "team:"
+
+func (o *OAuth) checkAccess(client *http.Client, restriction, login string) error {
+	if strings.HasPrefix(restriction, orgPrefix) {
+		org := strings.TrimPrefix(restriction, orgPrefix)
+		url := fmt.Sprintf("https://api.github.com/orgs/%s/members/%s", org, login)
+		resp, err := client.Get(url)
+		if err != nil {
+			return fmt.Errorf("can't get user organizations from github: %s", err)
+		}
+		// StatusNoContent means user is a member
+		// https://developer.github.com/v3/orgs/members/#check-membership
+		if resp.StatusCode != http.StatusNoContent {
+			return NoAccessError
+		}
+		return nil
+	}
+	if strings.HasPrefix(restriction, teamPrefix) {
+		team := strings.TrimPrefix(restriction, teamPrefix)
+		url := fmt.Sprintf("https://api.github.com/teams/%s/memberships/%s", team, login)
+		r, err := http.NewRequest("GET", url, nil)
+		// The Nested Teams API is currently available for developers to preview only
+		r.Header.Add("Accept", "application/vnd.github.hellcat-preview+json")
+		if err != nil {
+			return fmt.Errorf("can't create team request: %s", err)
+		}
+		resp, err := client.Do(r)
+		if err != nil {
+			return fmt.Errorf("can't get user organizations from github: %s", err)
+		}
+		// Only StatusOK means user is a member
+		// https://developer.github.com/v3/orgs/teams/#get-team-membership
+		if resp.StatusCode != http.StatusOK {
+			return NoAccessError
+		}
+		// Don't allow access to pending members
```

Comment:
Why do we need to forbid "pending" members? If we don't care about that, all code below this comment can be replaced by
```go
return nil
```

Reply:
because pending member isn't member